### Tidy Data Exercises

Do your work for this exercise in a jupyter notebook or python script named tidy_data. Save this work in your classification-exercises repo. Then add, commit, and push your changes.

In [2]:
import pandas as pd
import numpy as np
import seaborn as sns
from pydataset import data

1. Attendance Data

Load the attendance.csv file and calculate an attendance percentage for each student. One half day is worth 50% of a full day, and 10 tardies is equal to one absence.

You should end up with something like this:

    name
    Billy    0.5250
    Jane     0.6875
    John     0.9125
    Sally    0.7625
    
    Name: grade, dtype: float64

In [25]:
attendance = pd.read_csv('untidy-data/attendance.csv')
attendance.head()

Unnamed: 0 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0      Sally          P          T          T          H          P   
1       Jane          A          P          T          T          T   
2      Billy          A          T          A          A          H   
3       John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [26]:
attendance.shape

(4, 9)

In [27]:
# renaming student name column
attendance = attendance.rename(columns = {'Unnamed: 0': 'Name'})
attendance.head(2)

Name 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05 2018-01-06  \
0  Sally          P          T          T          H          P          A   
1   Jane          A          P          T          T          T          T   

  2018-01-07 2018-01-08  
0          T          T  
1          A          T

In [28]:
attendance = attendance.melt(id_vars = 'Name',
                            var_name = 'Date',
                            value_name = 'Attendance')
attendance

Name        Date Attendance
0   Sally  2018-01-01          P
1    Jane  2018-01-01          A
2   Billy  2018-01-01          A
3    John  2018-01-01          P
4   Sally  2018-01-02          T
5    Jane  2018-01-02          P
6   Billy  2018-01-02          T
7    John  2018-01-02          T
8   Sally  2018-01-03          T
9    Jane  2018-01-03          T
10  Billy  2018-01-03          A
11   John  2018-01-03          H
12  Sally  2018-01-04          H
13   Jane  2018-01-04          T
14  Billy  2018-01-04          A
15   John  2018-01-04          P
16  Sally  2018-01-05          P
17   Jane  2018-01-05          T
18  Billy  2018-01-05          H
19   John  2018-01-05          P
20  Sally  2018-01-06          A
21   Jane  2018-01-06          T
22  Billy  2018-01-06          T
23   John  2018-01-06          T
24  Sally  2018-01-07          T
25   Jane  2018-01-07          A
26  Billy  2018-01-07          P
27   John  2018-01-07          P
28  Sally  2018-01-08          T
29   Jane  2018-01-08          T
30  Billy  2018-01-08          T
31   John  2018-01-08          P

In [32]:
# map attendance codes to numeric values
attendance_code = {'A': 0, 'P':1, 'T':.9, 'H':.5,}

attendance['Attendance_Score'] = attendance['Attendance'].map(attendance_code)
attendance.head(2)

Name        Date Attendance  Attendance_Score
0  Sally  2018-01-01          P               1.0
1   Jane  2018-01-01          A               0.0

In [24]:
# drop columns and group by student name and average attendance score
attendance = attendance.drop(columns = ['Date', 'Attendance'])
attendance.groupby('Name').Attendance_Score.mean()

Name
Billy    0.5250
Jane     0.6875
John     0.9125
Sally    0.7625
Name: Attendance_Score, dtype: float64

2. Coffee Levels

 - Read the coffee_levels.csv file.

In [41]:
coffee = pd.read_csv('untidy-data/coffee_levels.csv')
coffee.head()

hour coffee_carafe  coffee_amount
0     8             x       0.816164
1     9             x       0.451018
2    10             x       0.843279
3    11             x       0.335533
4    12             x       0.898291

In [42]:
coffee.shape

(30, 3)

 - Transform the data so that each carafe is in it's own column.

In [43]:
# transforming the dataframe by splitting coffee_carafe into different columns
coffee = coffee.pivot(index = 'hour',
                     columns = 'coffee_carafe').reset_index()

In [44]:
coffee

hour coffee_amount                    
coffee_carafe                  x         y         z
0                8      0.816164  0.189297  0.999264
1                9      0.451018  0.521502  0.915990
2               10      0.843279  0.023163  0.144928
3               11      0.335533  0.235529  0.311495
4               12      0.898291  0.017009  0.771947
5               13      0.310711  0.997464  0.398520
6               14      0.507288  0.058361  0.864464
7               15      0.215043  0.144644  0.436364
8               16      0.183891  0.544676  0.280621
9               17      0.391560  0.594126  0.436677

In [39]:
coffee.shape

(10, 3)

 - Is this the best shape for the data?

It is definitely easier to read and understand

3. Cake Recipes

 - Read the cake_recipes.csv data. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.

In [71]:
cake_recipes = pd.read_csv('untidy-data/cake_recipes.csv')
cake_recipes.head()

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084

 - Tidy the data as necessary.

In [72]:
cake_recipes[['recipe:position', 'position']] = cake_recipes['recipe:position'].str.split(':', expand = True)
cake_recipes

recipe:position        225        250        275        300 position
0               a  61.738655  53.912627  74.414730  98.786784   bottom
1               a  51.709751  52.009735  68.576858  50.228470      top
2               b  57.095320  61.904369  61.196980  99.248541   bottom
3               b  82.455004  95.224151  98.594881  58.169349      top
4               c  96.470207  52.001358  92.893227  65.473084   bottom
5               c  71.306308  82.795477  92.098049  53.960273      top
6               d  52.799753  58.670419  51.747686  56.183110   bottom
7               d  96.873178  76.101363  59.571620  50.971626      top

In [73]:
# rename recipe:position  and position columns
cake_recipes = cake_recipes.rename(columns = {'recipe:position': 'recipe', 'position': 'rack_position'})
cake_recipes

recipe        225        250        275        300 rack_position
0      a  61.738655  53.912627  74.414730  98.786784        bottom
1      a  51.709751  52.009735  68.576858  50.228470           top
2      b  57.095320  61.904369  61.196980  99.248541        bottom
3      b  82.455004  95.224151  98.594881  58.169349           top
4      c  96.470207  52.001358  92.893227  65.473084        bottom
5      c  71.306308  82.795477  92.098049  53.960273           top
6      d  52.799753  58.670419  51.747686  56.183110        bottom
7      d  96.873178  76.101363  59.571620  50.971626           top

In [74]:
cake_recipes = cake_recipes.melt(id_vars = ['recipe', 'rack_position'],
                                var_name = 'oven_temperature',
                                value_name = 'rating')
cake_recipes.head(2)

recipe rack_position oven_temperature     rating
0      a        bottom              225  61.738655
1      a           top              225  51.709751

 - Which recipe, on average, is the best? recipe b

In [75]:
cake_recipes.groupby('recipe').rating.mean()

recipe
a    63.922201
b    76.736074
c    75.874748
d    62.864844
Name: rating, dtype: float64

 - Which oven temperature, on average, produces the best results? 275

In [76]:
cake_recipes.groupby('oven_temperature').rating.mean()

oven_temperature
225    71.306022
250    66.577437
275    74.886754
300    66.627655
Name: rating, dtype: float64

- Which combination of recipe, rack position, and temperature gives the best result? recipe b, bottom rack, 300 degrees

In [82]:
cake_recipes[cake_recipes.rating == cake_recipes.rating.max()]

recipe rack_position oven_temperature     rating
26      b        bottom              300  99.248541

In [86]:
cake_recipes.groupby(['oven_temperature', 'recipe', 'rack_position']).rating.mean()

oven_temperature  recipe  rack_position
225               a       bottom           61.738655
                          top              51.709751
                  b       bottom           57.095320
                          top              82.455004
                  c       bottom           96.470207
                          top              71.306308
                  d       bottom           52.799753
                          top              96.873178
250               a       bottom           53.912627
                          top              52.009735
                  b       bottom           61.904369
                          top              95.224151
                  c       bottom           52.001358
                          top              82.795477
                  d       bottom           58.670419
                          top              76.101363
275               a       bottom           74.414730
                          top              68.576858
      